src/lib/run.rs

```diff
@@ -1379,4 +1379,148 @@ mod test {
         assert_eq!(std::str::from_utf8(&i1s[0].seq).unwrap(), "GGTCAGAT"); // no masking
         assert_eq!(std::str::from_utf8(&u1s[0].seq).unwrap(), "ACGTAA"); // no masking
     }
+
+    #[rstest]
+    #[allow(clippy::too_many_lines)]
+    fn test_dual_index_hop_metrics() {
+        let dir = tempfile::tempdir().unwrap();
+        let fastqs = vec!["r1", "r2"]
+            .into_iter()
+            .map(|name| dir.path().join(format!("{}.fastq.gz", name)))
+            .collect_vec();
+
+        let read_structures = vec![
+            ReadStructure::from_str("3B+T").unwrap(), // 6bp UMI then template
+            ReadStructure::from_str("3B+T").unwrap(), // just template
+        ];
+
+        let r1 = Fq {
+            name: "q1",
+            bases: b"TTTCCC", //
+            quals: Some(b"IIICCC"),
+            ..Fq::default()
+        };
+        let r2 = Fq {
+            name: "q2",
+            bases: b"AAACCC", //
+            quals: Some(b"IIICCC"),
+            ..Fq::default()
+        };
+
+        for (fastq, read) in fastqs.iter().zip(vec![r1, r2].iter()) {
+            write_reads_to_file(std::iter::once(read.to_owned_record()), fastq);
+        }
+
+        let output = dir.path().join("output");
+        create_dir(&output).unwrap();
+
+        // Two barcodes for reads to hop
+        let metadata = dir.path().join("samples.csv");
+        let metadata_txt = r#"[Demux]
+[Data]
+Sample_ID,Index1_Sequence,Index2_Sequence
+s1,TTT,GGG
+s2,CCC,AAA"#;
+
+        std::fs::write(&metadata, metadata_txt).expect("Failed to write sample metadata.");
+
+        let opts = Opts {
+            fastqs,
+            output_dir: output.clone(),
+            sample_metadata: metadata.clone(),
+            read_structures,
+            output_types: "TB".to_owned(),
+            ..Opts::default()
+        };
+
+        let sample_sheet = SampleSheet::from_path(opts).unwrap();
+        run(sample_sheet.opts).unwrap();
+
+        // Check metrics
+        let hop_metrics = output.join("sample_barcode_hop_metrics.tsv");
+
+        let delim = DelimFile::default();
+        let hop_metrics: Vec<BarcodeCount> = delim.read_tsv(&hop_metrics).unwrap();
+        assert_eq!(hop_metrics.len(), 1);
+        assert_eq!(hop_metrics[0].barcode, "TTT+AAA");
```

Comment:
@nh13 @jiangweiyao This test demultiplexes a small example of dual indexed data where a single read hops a pair of sample barcodes. We check for that barcode in the output `barcode_hop_metrics.tsv`